apps/api/src/domain/models/shared/command-handlers/command-handler-base.ts

```diff
@@ -0,0 +1,84 @@
+import { Ack, ICommand, ICommandHandler } from '@coscrad/commands';
+import { buildSimpleValidationFunction } from '@coscrad/validation';
+import { Inject } from '@nestjs/common';
+import { InternalError, isInternalError } from '../../../../lib/errors/InternalError';
+import { RepositoryProvider } from '../../../../persistence/repositories/repository.provider';
+import { ResultOrError } from '../../../../types/ResultOrError';
+import { Valid } from '../../../domainModelValidators/Valid';
+import { IIdManager } from '../../../interfaces/id-manager.interface';
+import { InMemorySnapshot } from '../../../types/ResourceType';
+import InvalidCommandPayloadTypeError from '../common-command-errors/InvalidCommandPayloadTypeError';
+
+export abstract class CommandHandlerBase<TAggregate> implements ICommandHandler {
+    constructor(
+        protected readonly repositoryProvider: RepositoryProvider,
+        @Inject('ID_MANAGER') protected readonly idManager: IIdManager
+    ) {}
+
+    protected validateType(command: ICommand, commandType: string): Valid | InternalError {
+        // Validate command type
+        const payloadTypeErrors = buildSimpleValidationFunction(
+            Object.getPrototypeOf(command).constructor
+        )(command).map(
+            (simpleError) => new InternalError(`invalid payload type: ${simpleError.toString()}`)
+        );
+
+        if (payloadTypeErrors.length > 0) {
+            // TODO PAss through the command type
+            return new InvalidCommandPayloadTypeError(commandType, payloadTypeErrors);
+        }
+
+        return Valid;
+    }
+
+    abstract createOrFetchWriteContext(command: ICommand): Promise<ResultOrError<TAggregate>>;
+
+    abstract fetchRequiredExternalState(): Promise<InMemorySnapshot>;
+
+    // TODO Consider putting this on the instanc (e.g. an `applyCommand(type,payload)` method)
+    abstract actOnInstance(instance: TAggregate): ResultOrError<TAggregate>;
+
+    // TODO Put this on the Aggregate classes
+    abstract validateExternalState(
+        state: InMemorySnapshot,
+        instance: TAggregate
+    ): Valid | InternalError;
+
+    abstract persist(instance: TAggregate, command: ICommand): Promise<void>;
+
+    protected async validateAdditionalConstraints(_: ICommand): Promise<Valid | InternalError> {
+        return Valid;
+    }
+
+    async execute(command: ICommand, commandType: string): Promise<Ack | InternalError> {
+        const typeValidationResult = this.validateType(command, commandType);
```

Comment:
There must be a pattern that avoids spelling out the same error check over and over. Then again, this is isolated to one spot and it's probably not worth the effort.